src/sage/rings/polynomial/integer_valued_polynomials.py

```diff
@@ -0,0 +1,1125 @@
+                    in the binomial basis
+                """
+                real = self.realization_of()
+                return f"{real} in the {self._realization_name()} basis"
+
+            @cached_method
+            def one_basis(self):
+                r"""
+                Return the number 0, which index the unit of this algebra.
+
+                EXAMPLES::
+
+                    sage: A = IntegerValuedPolynomialRing(QQ).S()
+                    sage: A.one_basis()
+                    0
+                    sage: A.one()
+                    S[0]
+                """
+                return self.basis().keys()(0)
+
+            def gen(self):
+                r"""
+                Return the generator of this algebra.
+
+                EXAMPLES::
+
+                    sage: F = IntegerValuedPolynomialRing(ZZ).B()
+                    sage: F.gen()
+                    B[1]
+                """
+                return self.algebra_generators()[0]
+
+            @cached_method
+            def algebra_generators(self):
+                r"""
+                Return the generators of this algebra.
+
+                EXAMPLES::
+
+                    sage: A = IntegerValuedPolynomialRing(ZZ).S(); A
+                    Integer-Valued Polynomial Ring over Integer Ring
+                    in the shifted basis
+                    sage: A.algebra_generators()
+                    Family (S[1],)
+                """
+                NonNeg = self.basis().keys()
+                return Family([self.monomial(NonNeg(1))])
+
+            gens = algebra_generators
+
+        class ElementMethods:
+            def __call__(self, v):
+                """
+                Evaluation at some value ``v``
+
+                EXAMPLES::
+
+                     sage: F = IntegerValuedPolynomialRing(ZZ).S()
+                     sage: B = F.gen()
+                     sage: f = B**2+4*B+6
+                     sage: f(1/3)
+                     118/9
+
+                     sage: F = IntegerValuedPolynomialRing(ZZ).B()
+                     sage: B = F.gen()
+                     sage: f = B**2+4*B+6
+                     sage: f(1/3)
+                     67/9
+                """
+                return self.polynomial()(v)
+
+            def shift(self, j=1):
+                """
+                Shift all indices by `j`.
+
+                INPUT:
+
+                - `j` -- integer (default: 1)
+
+                In the binomial basis, the shift by 1 corresponds to
+                a summation operator from `0` to `x`.
+
+                EXAMPLES::
+
+                    sage: F = IntegerValuedPolynomialRing(ZZ).B()
+                    sage: B = F.gen()
+                    sage: (B+1).shift()
+                    B[1] + B[2]
+                    sage: (B+1).shift(3)
+                    B[3] + B[4]
+                """
+                A = self.parent()
+                return A._from_dict({i + j: c for i, c in self})
```

Comment:
You might want to make sure that the indices are in `self.basis().keys()` (which can be quickly accessed via `self._indices`). I could pass `1/2` or `4/2`, which might lead to subtle problems.

Reply:
done, not yet pushed